src/core/net/ip6_mpl.hpp

```diff
@@ -145,35 +148,217 @@ class OptionMpl: public OptionHeader
+
+private:
+    uint16_t mSeedId;
+    uint8_t  mSequence;
+    uint8_t  mLifetime;
+};
+
+
+/**
+ * This class represents MPL's Buffered Message entry.
+ *
+ */
+class MplBufferedMessageEntry
+{
+public:
+    /**
+     * This constructor creates an MPL buffered message entry instance.
+     *
+     */
+    MplBufferedMessageEntry(Mpl &aMpl, Ip6 &aIp6, TrickleTimer::Handler aHandler);
+
+    /**
+     * This method returns the pointer to the MPL Data Message.
+     *
+     * @returns The MPL Data Message pointer.
+     *
+     */
+    Message *GetMessage() const { return mMessage; }
+
+    /**
+     * This method sets the MPL Data Message.
+     *
+     * @param[in]  aMessage  The MPL Data Message pointer.
+     *
+     */
+    void SetMessage(Message *aMessage) { mMessage = aMessage; }
+
+    /**
+     * This method checks if entry is available to use.
+     *
+     * @return True MPL Buffered Message entry is free.
+     * @return False MPL Bufferred Messsage entry is occupied.
+     */
+    bool IsFree() { return mMessage == NULL; }
+
+    /**
+     * This method returns the reference to retransmission trickle timer.
+     *
+     * @returns The pointer to the retransmission trickle timer.
+     *
+     */
+    TrickleTimer &GetRetransmissionTimer() { return mRetransmissionTimer; }
+
+    /**
+     * This method returns the MPL Seed Id value.
+     *
+     * @returns The MPL Seed Id value.
+     *
+     */
+    uint16_t GetSeedId() const { return mSeedId; }
+
+    /**
+     * This method sets the MPL Seed Id value.
+     *
+     * @param[in]  aSeedId  The MPL Seed Id value.
+     *
+     */
+    void SetSeedId(uint16_t aSeedId) { mSeedId = aSeedId; }
+
+    /**
+     * This method returns the MPL Sequence value.
+     *
+     * @returns The MPL Sequence value.
+     *
+     */
+    uint8_t GetSequence() const { return mSequence; }
+
+    /**
+     * This method sets the MPL Sequence value.
+     *
+     * @param[in]  aSequence  The MPL Sequence value.
+     *
+     */
+    void SetSequence(uint8_t aSequence) { mSequence = aSequence; }
+
+    /**
+     * This method returns the number of already scheduled retransmissions.
+     *
+     * @returns The number of already scheduled retransmissions.
+     *
+     */
+    uint8_t GetTxCount() const { return mTxCount; }
+
+    /**
+     * This method sets the number of already scheduled retransmissions.
+     *
+     * @param[in]  aTxCount  The number of already scheduled retransmissions.
+     *
+     */
+    void SetTxCount(uint8_t aTxCount) { mTxCount = aTxCount; }
+
+    /**
+     * This method frees MPL Buffered Message entry.
+     *
+     */
+    void Free() { mMessage = NULL; }
+
+    /**
+     * This method returns reference to the MPL instance that entry is bound to.
+     *
+     * @returns The reference to the MPL instance.
+     */
+    Mpl &GetMpl() const { return mMpl; }
+
+private:
+    Mpl &mMpl;
+
+    Message *mMessage;
+    TrickleTimer mRetransmissionTimer;
```

Comment:
I know you're trying to deliver a spec-compliant MPL implementation.  At the same time, Thread only uses a very small subset of the MPL specification.  For example, Thread uses equal values for Imin and Imax, resulting in a fixed backoff window for each (re)transmission and does not enable Trickle suppression.

I would consider storing the state along with each message, as was done in #674.  This approach has several advantages:

1.  Reduces memory overhead by not maintaining a full TrickleTimer object.
2. Avoids having an independent limit on the number of buffered messages.
3. Avoids having fixed, statically allocated memory that is only used when there are buffered messages.
4. Avoids using the `new` operator.

Thoughts?